File: typescript/src/ancilis/testing/assertions.ts
/** Compliance assertion helpers for use in vitest/jest tests. */

import type { EvaluationResult } from "../engine/result.js";
import { ScanResult } from "./scan-result.js";

type ScanOrResult = ScanResult | EvaluationResult;

function toScanResult(target: ScanOrResult): ScanResult {
  if (target instanceof ScanResult) return target;
  return ScanResult.fromSingle(target);
}

function availableControls(result: ScanResult): string[] {
  if (result.evaluations.length === 0) return [];
  return result.evaluations[result.evaluations.length - 1]!.controlResults.map((cr) => cr.controlId);
}

function failingControls(result: ScanResult): string[] {
  const failing: string[] = [];
  for (const ev of result.evaluations) {
    for (const cr of ev.controlResults) {
      if (cr.result === "FAIL" || cr.result === "ERROR" || cr.result === "FLAG") {
        failing.push(`${cr.controlId}=${cr.result}: ${cr.detail}`);
      }
    }
  }
  return failing;
}

/**
 * Assert that the given control passed.
 *
 * Throws AssertionError with detailed context if the control did not pass.
 *
 * @example
 * assertControlPasses(scan, "PR-01");
 */
export function assertControlPasses(scan: ScanOrResult, controlId: string): void {
  const result = toScanResult(scan);
  const cr = result.getControlResult(controlId);
  if (cr === undefined) {
    throw new Error(
      `Control '${controlId}' was not evaluated. Available controls: ${availableControls(result).join(", ")}`,
    );
  }
  if (cr.result !== "PASS") {
    throw new Error(
      `Expected control '${controlId}' to PASS but got '${cr.result}'.\n  Detail: ${cr.detail}\n  Evidence: ${JSON.stringify(cr.evidenceData)}`,
    );
  }
}

/**
 * Assert that the given control failed (result is FAIL or ERROR).
 *
 * @example
 * assertControlFails(scan, "PR-01");
 */
export function assertControlFails(scan: ScanOrResult, controlId: string): void {
  const result = toScanResult(scan);
  const cr = result.getControlResult(controlId);
  if (cr === undefined) {
    throw new Error(
      `Control '${controlId}' was not evaluated. Available controls: ${availableControls(result).join(", ")}`,
    );
  }
  if (cr.result !== "FAIL" && cr.result !== "ERROR") {
    throw new Error(
      `Expected control '${controlId}' to FAIL but got '${cr.result}'.\n  Detail: ${cr.detail}\n  Evidence: ${JSON.stringify(cr.evidenceData)}`,
    );
  }
}

/**
 * Assert that the given control raised a FLAG.
 *
 * @example
 * assertControlFlags(scan, "DE-01");
 */
export function assertControlFlags(scan: ScanOrResult, controlId: string): void {
  const result = toScanResult(scan);
  const cr = result.getControlResult(controlId);
  if (cr === undefined) {
    throw new Error(
      `Control '${controlId}' was not evaluated. Available controls: ${availableControls(result).join(", ")}`,
    );
  }
  if (cr.result !== "FLAG") {
    throw new Error(
      `Expected control '${controlId}' to FLAG but got '${cr.result}'.\n  Detail: ${cr.detail}\n  Evidence: ${JSON.stringify(cr.evidenceData)}`,
    );
  }
}

/**
 * Assert that the overall posture score is above a threshold.
 *
 * Score is the pass rate across all scored controls (SKIP excluded).
 *
 * @param threshold - Float in [0.0, 1.0]. For example, 0.80 means 80% pass rate.
 *
 * @example
 * assertPostureAbove(scan, 0.80);
 */
export function assertPostureAbove(scan: ScanOrResult, threshold: number): void {
  const result = toScanResult(scan);
  const score = result.score;
  if (score < threshold) {
    throw new Error(
      `Posture score ${(score * 100).toFixed(1)}% is below required threshold ${(threshold * 100).toFixed(1)}%.\n  Failing controls: ${failingControls(result).join("; ")}`,
    );
  }
}

/**
 * Assert that the most recent evaluation decision is ALLOW.
 *
 * @example
 * assertDecisionAllows(scan);
 */
export function assertDecisionAllows(scan: ScanOrResult): void {
  const result = toScanResult(scan);
  const decision = result.decision();
  if (decision !== "ALLOW") {
    throw new Error(
      `Expected decision ALLOW but got '${decision}'.\n  Failing controls: ${failingControls(result).join("; ")}`,
    );
  }
}

/**
 * Assert that the most recent evaluation decision is BLOCK.
 *
 * @example
 * assertDecisionBlocks(scan);
 */
export function assertDecisionBlocks(scan: ScanOrResult): void {
  const result = toScanResult(scan);
  const decision = result.decision();
  if (decision !== "BLOCK") {
    throw new Error(`Expected decision BLOCK but got '${decision}'.`);
  }
}

File: python/src/ancilis/testing/assertions.py
"""Compliance assertion helpers for use in pytest tests."""

from __future__ import annotations

from ancilis.engine.result import EvaluationResult
from ancilis.testing.scan_result import ScanResult


def _to_scan_result(target: ScanResult | EvaluationResult) -> ScanResult:
    if isinstance(target, EvaluationResult):
        return ScanResult.from_single(target)
    return target


def assert_control_passes(
    scan: ScanResult | EvaluationResult,
    control_id: str,
) -> None:
    """Assert that the given control passed.

    Raises AssertionError with detailed context if the control did not pass.

    Usage::

        assert_control_passes(ancilis_scan, "PR-01")
    """
    result = _to_scan_result(scan)
    cr = result.get_control_result(control_id)
    if cr is None:
        raise AssertionError(
            f"Control '{control_id}' was not evaluated. "
            f"Available controls: {_available_controls(result)}"
        )
    if cr.result != "PASS":
        raise AssertionError(
            f"Expected control '{control_id}' to PASS but got '{cr.result}'.\n"
            f"  Detail: {cr.detail}\n"
            f"  Evidence: {cr.evidence_data}"
        )


def assert_control_fails(
    scan: ScanResult | EvaluationResult,
    control_id: str,
) -> None:
    """Assert that the given control failed (result is FAIL or ERROR).

    Raises AssertionError with detailed context if the control did not fail.

    Usage::

        assert_control_fails(ancilis_scan, "PR-01")
    """
    result = _to_scan_result(scan)
    cr = result.get_control_result(control_id)
    if cr is None:
        raise AssertionError(
            f"Control '{control_id}' was not evaluated. "
            f"Available controls: {_available_controls(result)}"
        )
    if cr.result not in ("FAIL", "ERROR"):
        raise AssertionError(
            f"Expected control '{control_id}' to FAIL but got '{cr.result}'.\n"
            f"  Detail: {cr.detail}\n"
            f"  Evidence: {cr.evidence_data}"
        )


def assert_control_flags(
    scan: ScanResult | EvaluationResult,
    control_id: str,
) -> None:
    """Assert that the given control raised a FLAG.

    Usage::

        assert_control_flags(ancilis_scan, "DE-01")
    """
    result = _to_scan_result(scan)
    cr = result.get_control_result(control_id)
    if cr is None:
        raise AssertionError(
            f"Control '{control_id}' was not evaluated. "
            f"Available controls: {_available_controls(result)}"
        )
    if cr.result != "FLAG":
        raise AssertionError(
            f"Expected control '{control_id}' to FLAG but got '{cr.result}'.\n"
            f"  Detail: {cr.detail}\n"
            f"  Evidence: {cr.evidence_data}"
        )


def assert_posture_above(
    scan: ScanResult | EvaluationResult,
    threshold: float,
) -> None:
    """Assert that the overall posture score is above a threshold.

    Score is the pass rate across all scored controls (SKIP excluded).

    Args:
        scan: ScanResult or EvaluationResult to evaluate.
        threshold: Float in [0.0, 1.0]. For example, 0.80 means 80% pass rate.

    Usage::

        assert_posture_above(ancilis_scan, 0.80)
    """
    result = _to_scan_result(scan)
    score = result.score
    if score < threshold:
        raise AssertionError(
            f"Posture score {score:.2%} is below required threshold {threshold:.2%}.\n"
            f"  Failing controls: {_failing_controls(result)}"
        )


def assert_decision_allows(scan: ScanResult | EvaluationResult) -> None:
    """Assert that the most recent evaluation decision is ALLOW."""
    result = _to_scan_result(scan)
    decision = result.decision()
    if decision != "ALLOW":
        raise AssertionError(
            f"Expected decision ALLOW but got '{decision}'.\n"
            f"  Failing controls: {_failing_controls(result)}"
        )


def assert_decision_blocks(scan: ScanResult | EvaluationResult) -> None:
    """Assert that the most recent evaluation decision is BLOCK."""
    result = _to_scan_result(scan)
    decision = result.decision()
    if decision != "BLOCK":
        raise AssertionError(
            f"Expected decision BLOCK but got '{decision}'."
        )


# --- Internal helpers ---

def _available_controls(result: ScanResult) -> list[str]:
    if not result.evaluations:
        return []
    return [cr.control_id for cr in result.evaluations[-1].control_results]


def _failing_controls(result: ScanResult) -> list[str]:
    failing = []
    for ev in result.evaluations:
        for cr in ev.control_results:
            if cr.result in ("FAIL", "ERROR", "FLAG"):
                failing.append(f"{cr.control_id}={cr.result}: {cr.detail}")
    return failing
